Add fallback to standard IPC in log forwarding

pvollan opened the pull request:
#### b6d93c4882a9cc178799e917e7464449995d07b3
<pre>
Add fallback to standard IPC in log forwarding
<a href="https://bugs.webkit.org/show_bug.cgi?id=292810">https://bugs.webkit.org/show_bug.cgi?id=292810</a>
<a href="https://rdar.apple.com/151044944">rdar://151044944</a>

Reviewed by Sihui Liu.

This is not a runtime fallback, but guarded by a compile flag. The default behavior
is to use streaming IPC. This behavior is being enabled on older versions on iOS,
in order to get testing of using the standard IPC feature in log forwarding.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/LogClient.cpp:
(WebKit::LogClient::LogClient):
(WebKit::LogClient::log):
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/generate-derived-log-sources.py:
(generate_messages_file):
(generate_log_client_declarations_file):
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::stopListeningForIPC):
(WebKit::LogStream::logOnBehalfOfWebContent):
(WebKit::LogStream::setup):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::create): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::setupLogStream):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::platformDestroy):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
(WebKit::WebProcessProxy::platformDestroy): Deleted.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::setupLogStream):

Canonical link: <a href="https://commits.webkit.org/295404@main">https://commits.webkit.org/295404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21ee3e5c6acf0a2d8eddfbf5bf31daa47047a8cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110189 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55652 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79708 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107979 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19523 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94740 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60015 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/104451 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12819 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55030 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97655 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112677 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103591 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32138 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23642 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32503 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90966 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88416 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22540 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11093 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27472 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32062 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37434 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/127895 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31855 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34985 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/ref-i31-eq.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35196 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33414 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.messages.in:
```diff
@@ -96,8 +96,13 @@ messages -> WebProcessProxy WantsDispatchMessage {
+#if ENABLE(STREAMING_IPC_IN_LOG_FORWARDING)
     SetupLogStream(uint32_t pid, IPC::StreamServerConnectionHandle serverConnection, WebKit::LogStreamIdentifier identifier) -> (IPC::Semaphore streamWakeUpSemaphore, IPC::Semaphore streamClientWaitSemaphore)
 #endif
+#if !ENABLE(STREAMING_IPC_IN_LOG_FORWARDING)
```
Do you mean to use `#else`? The nested conditions seem weird

pvollan replied:
IIRC, we cannot use '#else' in `messages.in` files.

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -414,6 +414,11 @@ WebProcessProxy::~WebProcessProxy()
     while (m_numberOfTimesSuddenTerminationWasDisabled-- > 0)
         WebCore::enableSuddenTermination();
 
+#if PLATFORM(COCOA) && !ENABLE(STREAMING_IPC_IN_LOG_FORWARDING)
```
I think this should be `ENABLE(LOGD_BLOCKING_IN_WEBCONTENT) && ENABLE(STREAMING_IPC_IN_LOG_FORWARDING)` to match `addMessageReceiver`.
Maybe we should move this in `platformDestroy()` in WebProcessProxyCocoa.mm too.

pvollan replied:
That's a good catch! I am moving this to `platformDestroy()` in the upcoming patch.

Thanks for reviewing!

pvollan committed:
Add fallback to standard IPC in log forwarding
https://bugs.webkit.org/show_bug.cgi?id=292810
rdar://151044944

Reviewed by Sihui Liu.

This is not a runtime fallback, but guarded by a compile flag. The default behavior
is to use streaming IPC. This behavior is being enabled on older versions on iOS,
in order to get testing of using the standard IPC feature in log forwarding.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/LogClient.cpp:
(WebKit::LogClient::LogClient):
(WebKit::LogClient::log):
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/generate-derived-log-sources.py:
(generate_messages_file):
(generate_log_client_declarations_file):
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::stopListeningForIPC):
(WebKit::LogStream::logOnBehalfOfWebContent):
(WebKit::LogStream::setup):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::create): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::setupLogStream):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::platformDestroy):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
(WebKit::WebProcessProxy::platformDestroy): Deleted.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::setupLogStream):

Canonical link: https://commits.webkit.org/295404@main
